synapse/daemon.py

```diff
@@ -326,7 +326,7 @@ async def _onDmonFini(self):
         if finis:
             await asyncio.wait(finis)
 
-        finis = [link.fini() for link in self.connectedlinks]
+        finis = [link.fini() for link in self.links]
```

Comment:
`finis = [link.fini() for link in self.links]` 
should be
`finis = [link.fini() for link in list(self.links)]`
Otherwise we'll throw a Runtime Error due to the set size changing during iteration, because the fini handler for the link removes it from the set.